compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -283,86 +307,156 @@ generateTemplateInstance env typeCon typeParams externPkgId =
-                                                                            (T.unpack $
-                                                                             T.intercalate
-                                                                                 "." $
-                                                                             LF.unTypeConName
-                                                                                 templName) <>
-                                                                            ":" <>
-                                                                            T.unpack
-                                                                                funName))))
-                                                      ]
-                                                , grhssLocalBinds =
-                                                      noLoc emptyLocalBinds
-                                                }
-                                      }
-                                ]
-                      , mg_origin = Generated
+
+-- | Generate a single choice instance for a given templateDT/choice
+generateChoiceInstance ::
+       Env
+    -> LF.PackageId
+    -> LF.DefDataType
+    -> LF.TemplateChoice
+    -> HsDecl GhcPs
+generateChoiceInstance env externPkgId templateDT choice =
+    InstD noExt $
+    ClsInstD
+        noExt
+        ClsInstDecl
+            { cid_ext = noExt
+            , cid_poly_ty =
+                  HsIB
+                      { hsib_ext = noExt
+                      , hsib_body = body
                       }
-            , fun_co_fn = WpHole
-            , fun_tick = []
+            , cid_binds = listToBag $ map classMethodStub methodMapping
+            , cid_sigs = []
+            , cid_tyfam_insts = []
+            , cid_datafam_insts = []
+            , cid_overlap_mode = Nothing
             }
+  where
+
+    app :: LHsType GhcPs -> LHsType GhcPs -> LHsType GhcPs
+    app t1 t2 = noLoc $ HsAppTy noExt t1 t2
+
+    body :: LHsType GhcPs =
+      choiceClass `app` arg1 `app` arg2 `app` arg3
+
+    choiceClass :: LHsType GhcPs =
+        noLoc $
+        HsTyVar noExt NotPromoted $
+        noLoc $
+        mkRdrQual (mkModuleName "DA.Internal.Template") $
+        mkOccName varName "Choice" :: LHsType GhcPs
+
+    arg1 :: LHsType GhcPs =
+      noLoc $ convType env lfTemplateType
+
+    arg2 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceType
+
+    arg3 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceReturnType
+
+    moduleNameStr = T.unpack $ LF.moduleNameString $ LF.moduleName $ envMod env
+    moduleName0 =
+        LF.ModuleName $
+        map T.pack $
+        splitOn "." moduleNameStr
+
+    lfTemplateType :: LF.Type =
+        LF.mkTApps
+          (LF.TCon (LF.Qualified LF.PRSelf moduleName0 dataTypeCon))
+          (map (LF.TVar . fst) dataParams)
+
+    lfChoiceType :: LF.Type =
```

Comment:
This isn’t quite right:
1. The choice type does not need to be defined in the same module as the template. This matters both for `Archive` (is that why you filtered it out above?) and for generic templates where the monomorphized version will live in the module where you create the instance while the choice type will live in the module where you define the generic template.
2. In DAML-LF the name of the choice type constructor must not match the choice name. There is no good way to represent this in DAML and we never generate such code so I would suggest to error out in that case.

So in summary:
* Validate that the choice name and the type constructor name of the choice type match
* Use the choice argument type instead of inventing the choice type based on the choice name.

Reply:
Yes that's why I filtered `Archive`.
I was assuming that for a well-formed DAML the names must match.
If not, how do we discover from a `TemplateChoice` the corresponding `TypeConName`?

Reply:
In DAML the names match. In DAML-LF the definition of a template choice includes the name as well as the type as separate fields and no relationship between the two is required.

Reply:
The type being the 2nd element of the pair `chcArgBinder`, right?

Reply:
There are a couple of different restrictions at different levels so let me try to make them explicit (also responding to the comment below on why we probably don’t want to support mismatches between the choice name and the choice type constructor name).

1. If you define a template in DAML, we will generate a record type for each choice with the same name as the choice. There are two somewhat special cases here:
    1. The `Archive` record is defined in `daml-stdlib` so there is only one for all templates.
    2. For generic templates, we only defined one choice record that might have type parameters instead of defining one per instance.

 However, even for these cases the choice name is the same as the type constructor name.
2. The `Choice` typeclass in DAML requires the pair `(templateType, choiceArgType)` to uniquely identify a choice on a given template. In particular, that means that we cannot represent two choices with the same argument type in DAML. It does _not_ require that `choiceArgType` is a record and supporting variants should work out nicely and not be confusing. The `Choice` typeclass does also _not_ require that the type constructor name  and the choice name match. However, if that is not the case things get really confusing. For an extreme example consider a choice `Foo` which takes a type called `Bar` as the argument and a choice called `Bar` which takes a type called `Foo` as the argument. To call the `Foo` choice you would now do something like `exercise cid Bar` in DAML which at least to me isn’t intuitive at all. Furthermore, while you could support this case it falls apart if you have a template with two choices (with different names) that take the same choice argument type. Therefore, the most sensible option (at least initially) seems to me to reject choices that have different choice and type constructor names.
3. In DAML-LF the only restriction is that the choice names must unique per template. It is possible to use the same choice argument type for multiple choices, it is possible to use non-record types (e.g., even `()` which we should also reject). As mentioned above the `Choice` typeclass above does not allow us to express this in full generality so we necessarily have to reject some cases and we probably want to reject slightly more than strictly necessary to provide a reasonable UX.

Reply:
Fixed code to no longer assume there is a datatype named the same as the choice